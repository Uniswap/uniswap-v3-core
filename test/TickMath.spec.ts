import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { TickMathTest } from '../typechain/TickMathTest'
import { expect } from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'
import { encodePriceSqrt } from './shared/utilities'

const MIN_TICK = -887272
const MAX_TICK = 887272

const MIN_SQRT_RATIO = BigNumber.from('4295128739')
const MAX_SQRT_RATIO = BigNumber.from('1461446703485210103287273052203988822378723970342')

describe('TickMath', () => {
  let tickMath: TickMathTest

  before('deploy TickMathTest', async () => {
    const factory = await ethers.getContractFactory('TickMathTest')
    tickMath = (await factory.deploy()) as TickMathTest
  })

  describe('#getSqrtRatioAtTick', () => {
    it('throws for too low', async () => {
      await expect(tickMath.getSqrtRatioAtTick(MIN_TICK - 1)).to.be.revertedWith('T')
    })

    it('throws for too low', async () => {
      await expect(tickMath.getSqrtRatioAtTick(MAX_TICK + 1)).to.be.revertedWith('T')
    })

    it('min tick', async () => {
      expect(await tickMath.getSqrtRatioAtTick(MIN_TICK)).to.eq('4295128739')
    })

    it('min tick +1', async () => {
      expect(await tickMath.getSqrtRatioAtTick(MIN_TICK + 1)).to.eq('4295343490')
    })

    it('max tick - 1', async () => {
      expect(await tickMath.getSqrtRatioAtTick(MAX_TICK - 1)).to.eq('1461373636630004318706518188784493106690254656249')
    })

    it('min tick ratio is less than js implementation', async () => {
      expect(await tickMath.getSqrtRatioAtTick(MIN_TICK)).to.be.lt(encodePriceSqrt(1, BigNumber.from(2).pow(127)))
    })

    it('max tick ratio is greater than js implementation', async () => {
      expect(await tickMath.getSqrtRatioAtTick(MAX_TICK)).to.be.gt(encodePriceSqrt(BigNumber.from(2).pow(127), 1))
    })

    it('max tick', async () => {
      expect(await tickMath.getSqrtRatioAtTick(MAX_TICK)).to.eq('1461446703485210103287273052203988822378723970342')
    })

    for (const absTick of [
      50,
      100,
      250,
      500,
      1_000,
      2_500,
      3_000,
      4_000,
      5_000,
      50_000,
      150_000,
      250_000,
      500_000,
      738_203,
    ]) {
      for (const tick of [-absTick, absTick]) {
        describe(`tick ${tick}`, () => {
          it('result', async () => {
            expect((await tickMath.getSqrtRatioAtTick(tick)).toString()).to.matchSnapshot()
          })
          it('gas', async () => {
            await snapshotGasCost(tickMath.getGasCostOfGetSqrtRatioAtTick(tick))
          })
        })
      }
    }
  })

  describe('#MIN_SQRT_RATIO', async () => {
    it('equals #getSqrtRatioAtTick(MIN_TICK)', async () => {
      const min = await tickMath.getSqrtRatioAtTick(MIN_TICK)
      expect(min).to.eq(await tickMath.MIN_SQRT_RATIO())
      expect(min).to.eq(MIN_SQRT_RATIO)
    })
  })

  describe('#MAX_SQRT_RATIO', async () => {
    it('equals #getSqrtRatioAtTick(MAX_TICK)', async () => {
      const max = await tickMath.getSqrtRatioAtTick(MAX_TICK)
      expect(max).to.eq(await tickMath.MAX_SQRT_RATIO())
      expect(max).to.eq(MAX_SQRT_RATIO)
    })
  })

  describe('#getTickAtSqrtRatio', () => {
    it('throws for too low', async () => {
      await expect(tickMath.getTickAtSqrtRatio(MIN_SQRT_RATIO.sub(1))).to.be.revertedWith('R')
    })

    it('throws for too high', async () => {
      await expect(tickMath.getTickAtSqrtRatio(BigNumber.from(MAX_SQRT_RATIO))).to.be.revertedWith('R')
    })

    it('ratio of min tick', async () => {
      expect(await tickMath.getTickAtSqrtRatio(MIN_SQRT_RATIO)).to.eq(MIN_TICK)
    })
    it('ratio of min tick + 1', async () => {
      expect(await tickMath.getTickAtSqrtRatio('4295343490')).to.eq(MIN_TICK + 1)
    })
    it('ratio of max tick - 1', async () => {
      expect(await tickMath.getTickAtSqrtRatio('1461373636630004318706518188784493106690254656249')).to.eq(MAX_TICK - 1)
    })
    it('ratio closest to max tick', async () => {
      expect(await tickMath.getTickAtSqrtRatio(MAX_SQRT_RATIO.sub(1))).to.eq(MAX_TICK - 1)
    })

    for (const ratio of [MIN_SQRT_RATIO, MAX_SQRT_RATIO.sub(1)]) {
      describe(`ratio ${ratio}`, () => {
        it('result', async () => {
          expect((await tickMath.getTickAtSqrtRatio(ratio)).toString()).to.matchSnapshot()
        })
        it('gas', async () => {
          await snapshotGasCost(tickMath.getGasCostOfGetTickAtSqrtRatio(ratio))
        })
      })
    }
  })
})
