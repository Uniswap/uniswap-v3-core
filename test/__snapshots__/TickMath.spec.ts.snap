// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickMath #getSqrtRatioAtTick tick -50 gas 1`] = `829`;

exports[`TickMath #getSqrtRatioAtTick tick -50 result 1`] = `"79030349367926598376800521322"`;

exports[`TickMath #getSqrtRatioAtTick tick -100 gas 1`] = `829`;

exports[`TickMath #getSqrtRatioAtTick tick -100 result 1`] = `"78833030112140176575862854579"`;

exports[`TickMath #getSqrtRatioAtTick tick -250 gas 1`] = `871`;

exports[`TickMath #getSqrtRatioAtTick tick -250 result 1`] = `"78244023372248365697264290337"`;

exports[`TickMath #getSqrtRatioAtTick tick -500 gas 1`] = `871`;

exports[`TickMath #getSqrtRatioAtTick tick -500 result 1`] = `"77272108795590369356373805297"`;

exports[`TickMath #getSqrtRatioAtTick tick -1000 gas 1`] = `871`;

exports[`TickMath #getSqrtRatioAtTick tick -1000 result 1`] = `"75364347830767020784054125655"`;

exports[`TickMath #getSqrtRatioAtTick tick -2500 gas 1`] = `857`;

exports[`TickMath #getSqrtRatioAtTick tick -2500 result 1`] = `"69919044979842180277688105136"`;

exports[`TickMath #getSqrtRatioAtTick tick -3000 gas 1`] = `885`;

exports[`TickMath #getSqrtRatioAtTick tick -3000 result 1`] = `"68192822843687888778582228483"`;

exports[`TickMath #getSqrtRatioAtTick tick -4000 gas 1`] = `871`;

exports[`TickMath #getSqrtRatioAtTick tick -4000 result 1`] = `"64867181785621769311890333195"`;

exports[`TickMath #getSqrtRatioAtTick tick -5000 gas 1`] = `857`;

exports[`TickMath #getSqrtRatioAtTick tick -5000 result 1`] = `"61703726247759831737814779831"`;

exports[`TickMath #getSqrtRatioAtTick tick -50000 gas 1`] = `871`;

exports[`TickMath #getSqrtRatioAtTick tick -50000 result 1`] = `"6504256538020985011912221507"`;

exports[`TickMath #getSqrtRatioAtTick tick -150000 gas 1`] = `899`;

exports[`TickMath #getSqrtRatioAtTick tick -150000 result 1`] = `"43836292794701720435367485"`;

exports[`TickMath #getSqrtRatioAtTick tick -250000 gas 1`] = `885`;

exports[`TickMath #getSqrtRatioAtTick tick -250000 result 1`] = `"295440463448801648376846"`;

exports[`TickMath #getSqrtRatioAtTick tick -500000 gas 1`] = `885`;

exports[`TickMath #getSqrtRatioAtTick tick -500000 result 1`] = `"1101692437043807371"`;

exports[`TickMath #getSqrtRatioAtTick tick -738203 gas 1`] = `911`;

exports[`TickMath #getSqrtRatioAtTick tick -738203 result 1`] = `"7409801140451"`;

exports[`TickMath #getSqrtRatioAtTick tick 50 gas 1`] = `869`;

exports[`TickMath #getSqrtRatioAtTick tick 50 result 1`] = `"79426470787362580746886972461"`;

exports[`TickMath #getSqrtRatioAtTick tick 100 gas 1`] = `869`;

exports[`TickMath #getSqrtRatioAtTick tick 100 result 1`] = `"79625275426524748796330556128"`;

exports[`TickMath #getSqrtRatioAtTick tick 250 gas 1`] = `911`;

exports[`TickMath #getSqrtRatioAtTick tick 250 result 1`] = `"80224679980005306637834519095"`;

exports[`TickMath #getSqrtRatioAtTick tick 500 gas 1`] = `911`;

exports[`TickMath #getSqrtRatioAtTick tick 500 result 1`] = `"81233731461783161732293370115"`;

exports[`TickMath #getSqrtRatioAtTick tick 1000 gas 1`] = `911`;

exports[`TickMath #getSqrtRatioAtTick tick 1000 result 1`] = `"83290069058676223003182343270"`;

exports[`TickMath #getSqrtRatioAtTick tick 2500 gas 1`] = `897`;

exports[`TickMath #getSqrtRatioAtTick tick 2500 result 1`] = `"89776708723587163891445672585"`;

exports[`TickMath #getSqrtRatioAtTick tick 3000 gas 1`] = `925`;

exports[`TickMath #getSqrtRatioAtTick tick 3000 result 1`] = `"92049301871182272007977902845"`;

exports[`TickMath #getSqrtRatioAtTick tick 4000 gas 1`] = `911`;

exports[`TickMath #getSqrtRatioAtTick tick 4000 result 1`] = `"96768528593268422080558758223"`;

exports[`TickMath #getSqrtRatioAtTick tick 5000 gas 1`] = `897`;

exports[`TickMath #getSqrtRatioAtTick tick 5000 result 1`] = `"101729702841318637793976746270"`;

exports[`TickMath #getSqrtRatioAtTick tick 50000 gas 1`] = `911`;

exports[`TickMath #getSqrtRatioAtTick tick 50000 result 1`] = `"965075977353221155028623082916"`;

exports[`TickMath #getSqrtRatioAtTick tick 150000 gas 1`] = `939`;

exports[`TickMath #getSqrtRatioAtTick tick 150000 result 1`] = `"143194173941309278083010301478497"`;

exports[`TickMath #getSqrtRatioAtTick tick 250000 gas 1`] = `925`;

exports[`TickMath #getSqrtRatioAtTick tick 250000 result 1`] = `"21246587762933397357449903968194344"`;

exports[`TickMath #getSqrtRatioAtTick tick 500000 gas 1`] = `925`;

exports[`TickMath #getSqrtRatioAtTick tick 500000 result 1`] = `"5697689776495288729098254600827762987878"`;

exports[`TickMath #getSqrtRatioAtTick tick 738203 gas 1`] = `951`;

exports[`TickMath #getSqrtRatioAtTick tick 738203 result 1`] = `"847134979253254120489401328389043031315994541"`;

exports[`TickMath #getTickAtSqrtRatio ratio 4295128739 gas 1`] = `2389`;

exports[`TickMath #getTickAtSqrtRatio ratio 4295128739 result 1`] = `"-887272"`;

exports[`TickMath #getTickAtSqrtRatio ratio 1461446703485210103287273052203988822378723970341 gas 1`] = `2449`;

exports[`TickMath #getTickAtSqrtRatio ratio 1461446703485210103287273052203988822378723970341 result 1`] = `"887271"`;
